### DISCUSSION TITLE
The evolution of superquadric modeling and grasping

### DISCUSSION THREAD

**Discussion by pattacini** [robotology/community#471] [Research news]
We recently extended the superquadric modeling and grasping pipeline to deal with more complex objects!

Our robots are now able to reconstruct 3D models and to compute grasp candidates for a large number of objects:

![giphy](https://user-images.githubusercontent.com/3738070/103892985-c4d92080-50ec-11eb-9e3a-ed639d478575.gif)

## Technology
We use **nonlinear constrained optimization** both to fit the multi-superquadric model to the partial point cloud of the object and to obtain the grasping poses.

## Perks
- The multi-superquadric model is the result of a splitting and merging procedure.
- When computing a grasp candidate for a specific portion of the object the other portions are considered as obstacles to avoid object penetration with the robot hand.

## Outlook
- We're going to test the multi-superquadric approach on tool grasping with the iCub.
- We're planning to use our modeling and grasping strategy to guide RL training on our robots.

## Code
A new and efficient implementation of our (single and multiple) superquadric modeling and grasping algorithms is now available as a library: https://github.com/robotology/superquadric-lib.

## Publications
A new publication about multi-superquadric modeling and grasping is coming soon!

For now, you can have a look at the other papers explaining the evolution of our method:

- G. Vezzani, U. Pattacini, and L. Natale, [A grasping approach based on superquadric models](https://ieeexplore.ieee.org/abstract/document/7989187), ICRA 2017.
- G. Vezzani, U. Pattacini, G. Pasquale, and L. Natale, [Improving Superquadric Modeling and Grasping with Prior on Object Shapes](https://ieeexplore.ieee.org/abstract/document/8463161), ICRA 2018.

Credits go to @giuliavezzani

**Comment by jimingda** [comment]
Hi, what i want to do is to grasp the wheel by icub hands, so i find here, can i use this tool in my application?
![image](https://github.com/robotology/community/assets/7019692/13bbb5a1-cdff-4e13-9cc7-c7c602557159)

**Reply by jimingda** [reply]
Maybe the grasp in red ball demo is more helpful to me. I will refer to that first.

**Reply by pattacini** [reply]
Given that you're equipped with a perfect model of the object of interest, I suggest employing it and carrying out ad-hoc inverse kinematics, specifically customized to provide grasp candidates for the steering wheel.

**Reply by jimingda** [reply]
Thanks@pattacini 
Are there any tutorial of inverse kinematics or forward kinematics about fingers like thumb index and middle finger. Currently i have followed the air-hocky project and generated the trajectory of hand positions and poses on both arms, but still haven't control the fingers.

**Reply by pattacini** [reply]
Hi @jimingda 

As far as I know, there's no specific tutorial for precision grasp. We did some work in the past on 3-fingers grasping[^1], although it's definitely no longer maintained[^2]. 

That said, I wouldn't go for precision-grasp in your context, but rather for power-grasp as you need to exert a non-negligible amount of force to steer the wheel.

A tutorial on power-grasp based on the cardinal points method can be found at https://github.com/robotology/icub-gazebo-grasping-sandbox.

[^1]: [Three-Finger Precision Grasp on Incomplete 3D Point Clouds](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=6907648).
[^2]: Code is still available at https://github.com/robotology/grasp anyway.
